sys/shell/cmds/udptty.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2023 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @ingroup     sys_shell_commands
+ * @{
+ *
+ * @file
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "shell.h"
+#include "stdio_base.h"
+#include "net/sock/async.h"
+#include "net/sock/udp.h"
+#include "net/utils.h"
+#include "ztimer.h"
+
+extern int readline(char *buf, size_t size);
+
+static void _sock_cb(sock_udp_t *sock, sock_async_flags_t flags, void *arg)
+{
+    (void)arg;
+    if ((flags & SOCK_ASYNC_MSG_RECV) == 0) {
+        return;
+    }
+
+    void *data, *ctx = NULL;
+
+    int res;
+    while ((res = sock_udp_recv_buf(sock, &data, &ctx, 0, NULL)) > 0) {
+        stdio_write(data, res);
+    }
+}
+
+static int _udptty_cmd(int argc, char **argv)
+{
+    if (argc < 2) {
+        goto usage;
+    }
+
+    netif_t *netif = NULL;
+    sock_udp_ep_t remote = SOCK_IPV6_EP_ANY;
+    sock_udp_t sock;
+    int res;
+
+    const sock_udp_ep_t local = {
+        .family = AF_INET6,
+        .netif = SOCK_ADDR_ANY_NETIF,
+    };
+
+    if (netutils_get_ipv6((ipv6_addr_t *)&remote.addr, &netif, argv[1])) {
+        goto usage;
+    }
+
+    if (netif) {
+        remote.netif = netif_get_id(netif);
+    }
+    remote.port = atoi(argv[2]);
+
+    res = sock_udp_create(&sock, &local, &remote, 0);
+    if (res) {
+        printf("can't create socket: %d\n", res);
+        return res;
+    }
+
+    sock_udp_set_cb(&sock, _sock_cb, NULL);
+
+    do {
+        char buf[64];
+        res = readline(buf, sizeof(buf));
+
+        if (res > 0) {
+            buf[res] = '\n';
+            sock_udp_send(&sock, buf, res + 1, NULL);
+        }
+
+#ifdef CPU_NATIVE
+        /* readline() on native blocks all other threads - see #19002 */
+        ztimer_sleep(ZTIMER_MSEC, 100);
+#endif
+    } while (res != EOF);
+
+    /* send disconnect */
+    const char eof = EOF;
+    sock_udp_send(&sock, &eof, sizeof(eof), NULL);
+    sock_udp_close(&sock);
```

Comment:
`stdio_udp` doesn't seem to have special handling for `EOF`. As far as I understand, it will just replace the `remote` to send to whenever a datagram came in from a new endpoint.

Given the lack of reliable transfer in UDP, I wonder how helpful adding handling of `EOF` to `stdio_udp` would actually be. IMO it would be fine to just stop sending just like netcat does.

Reply:
The shell handles `EOF` but I agree that it would be better if `stdio_udp` did so.
When using `nc` I have to do `Ctrl-V` `Ctrl-D` to end the session, but it's important so that `stdio_udp` stops sending all stdio over the wire (possibly congesting the network) when the remote debug session has ended.

Reply:
#19738 will now handle this.